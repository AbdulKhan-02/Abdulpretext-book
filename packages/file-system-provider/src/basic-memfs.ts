export * from "./lib/basic-memfs";
