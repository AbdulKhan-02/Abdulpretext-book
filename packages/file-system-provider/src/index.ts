export * from "./lib/basic-memfs";
export { MemFS } from "./lib/memfs";
export { BasicMemFS } from "./lib/basic-memfs";
