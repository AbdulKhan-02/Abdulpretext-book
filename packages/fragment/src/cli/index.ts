#!/usr/bin/env node
import yargs from "yargs";
import { hideBin } from "yargs/helpers";
import fs from "node:fs/promises";
import { getTemplateName } from "../lib/get-template-name";
import { fragmentToXast } from "../lib/fragment-to-pretext";
import { toXml } from "xast-util-to-xml";
import { extractFragmentFromHtml } from "../lib/extract-fragment-from-html";

const yargsObj = yargs(hideBin(process.argv))
    .usage(
        `$0 --template-file ./template.ptx --fragment-file ./fragment.xml`.trim()
    )
    .options({
        "template-file": {
            type: "string",
            description:
                "Path to the template file. This file should have a <fragment /> element where you want the substitution to occur",
            requiresArg: true,
        },
        template: {
            type: "string",
            description:
                "Template as a string. This string should have a <fragment /> element where you want the substitution to occur",
            requiresArg: true,
        },
        "template-name": {
            type: "boolean",
            description:
                "Extract the name of the template requested by the fragment.",
        },
        "fragment-file": {
            type: "string",
            description:
                "Path to the fragment file. This file should have a <fragment /> element as the root node.",
            requiresArg: true,
        },
        fragment: {
            type: "string",
            description:
                "Fragment as a string. This string should have a <fragment /> element as the root node.",
            requiresArg: true,
        },
        "extract-from-html-file": {
            type: "string",
            description:
                "Extract the rendered fragment from the specified file.",
            requiresArg: true,
        },
        "extract-from-html": {
            type: "string",
            description:
                "Extract the rendered fragment from the specified string.",
            requiresArg: true,
        },
        "pretty-print": {
            type: "boolean",
            description:
                "Whether to pretty-print fragments extracted from HTML",
        },
        out: {
            type: "string",
            alias: "o",
            description: "Path to the file where the result will be saved.",
            requiresArg: true,
        },
    })
    .help();
const options = yargsObj.argv;

type Options = Awaited<typeof options>;

const cache: { template?: string; fragment?: string; html?: string } = {};

/**
 * Cached function to get the contents of the _fragment_ as specified
 * by command-line options
 */
async function getFragment(options: Options) {
    if (cache.fragment) {
        return cache.fragment;
    }
    return (
        options.fragment ||
        (options.fragmentFile &&
            (await fs.readFile(options.fragmentFile, { encoding: "utf-8" })))
    );
}

/**
 * Cached function to get the contents of the _template_ as specified
 * by command-line options
 */
async function getTemplate(options: Options) {
    if (cache.fragment) {
        return cache.fragment;
    }
    return (
        options.template ||
        (options.templateFile &&
            (await fs.readFile(options.templateFile, { encoding: "utf-8" })))
    );
}

/**
 * Cached function to extract the contents of the rendered fragment from the _html_ as specified
 * by command-line options
 */
async function getHtml(options: Options) {
    if (cache.html) {
        return cache.html;
    }
    return (
        options.extractFromHtml ||
        (options.extractFromHtmlFile &&
            (await fs.readFile(options.extractFromHtmlFile, {
                encoding: "utf-8",
            })))
    );
}

async function getTemplateNameFromFragment(options: Options) {
    const fragment = await getFragment(options);
    if (!fragment) {
        throw new Error(
            `Must supply a fragment to extract its requested template's name.`
        );
    }
    return getTemplateName(fragment);
}

async function evaluateTemplate(options: Options) {
    const fragment = await getFragment(options);
    const template = await getTemplate(options);
    if (!fragment) {
        throw new Error(
            `You must supply a fragment string or file to evaluate a template`
        );
    }
    if (!template) {
        throw new Error(
            `You must supply a template string or file to evaluate a template`
        );
    }

    // On the command line, we just assume that the supplied template is the requested template
    const templateName = getTemplateName(fragment);
    const renderedAst = fragmentToXast(fragment, { [templateName]: template });
    return toXml(renderedAst);
}

async function extractFromHtml(options: Options) {
    const html = await getHtml(options);
    if (!html) {
        throw new Error(
            `You must supply an html string or file to extract a fragment from`
        );
    }
    return extractFragmentFromHtml(html, { prettyPrint: options.prettyPrint });
}

/**
 * Write to a file or log to stdout the contents of `str`, depending on
 * whether `--out` was specified in `options`.
 */
async function writeOrLog(str: string, options: Options) {
    if (!options.out) {
        console.log(str);
        return;
    }
    await fs.writeFile(options.out, str, { encoding: "utf-8", flag: "w" });
}

async function main() {
    const opts = await options;

    if (opts.templateName) {
        console.log(await getTemplateNameFromFragment(opts));
        return;
    }

    if (opts.fragment || opts.fragmentFile) {
        const rendered = await evaluateTemplate(opts);
        await writeOrLog(rendered, opts);
        return;
    }

    if (opts.extractFromHtml || opts.extractFromHtmlFile) {
        const extracted = await extractFromHtml(opts);
        await writeOrLog(extracted, opts);
        return;
    }

    // If no arguments were given, show the help
    yargsObj.showHelp();
}

main();
