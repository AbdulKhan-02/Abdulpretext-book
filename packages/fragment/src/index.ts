export { CssSelectorParser } from "./parser";
import { toXml } from "xast-util-to-xml";
import { fragmentToXast } from "./lib/fragment-to-pretext";
export { jestToMatchFragment } from "./lib/jest-render-as-fragment";
export { getTemplateName } from "./lib/get-template-name";
export { extractFragmentFromHtml } from "./lib/extract-fragment-from-html";
export { normalizeFragmentHtml } from "./lib/normalize-fragment-html";

export function fragmentToPretext(
    source: string,
    templates: Record<string, string>
): string {
    const rendered = fragmentToXast(source, templates);
    return toXml(rendered);
}
