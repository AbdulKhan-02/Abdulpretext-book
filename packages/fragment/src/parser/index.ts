// This file needs to be here because typescript does not know how to use the transpiler
// to directly load Pegjs grammars.
// @ts-nocheck
import _CssSelectorParser from "../grammars/css-selector.peggy";
import type { ParseFunction } from "../grammars/peggy-types";

type PegParser = {
    parse: ParseFunction;
    SyntaxError: (
        message: string,
        expected: string,
        found: unknown,
        location: unknown
    ) => unknown;
};

const CssSelectorParser = _CssSelectorParser as PegParser;

export { CssSelectorParser };
