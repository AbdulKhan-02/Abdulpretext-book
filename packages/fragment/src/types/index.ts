import {
    Element as XastElement,
    Text as XastText,
    Root as XastRoot,
} from "xast";
import { XastNode } from "../utils/xast";

export { XastElement, XastText, XastRoot };

export type Fragment = {
    content: XastNode | XastNode[];
    template: string;
    selector: string;
};
