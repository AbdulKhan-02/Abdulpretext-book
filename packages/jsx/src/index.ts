import { normalizePretextXml } from "./relax-ng/normalize-pretext-xml";

export { normalizePretextXml };
