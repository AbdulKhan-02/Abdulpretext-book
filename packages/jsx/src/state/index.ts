import Slugger from "github-slugger";
import { toXml } from "xast-util-to-xml";
import { JsonGrammar } from "../utils/relax-ng/types";
import { elmMatcher, isElement, onlyElementsAndText } from "../utils/tools";
import {
    ProcessContentFunc,
    visit,
    XastNode,
    XastElement,
    XastRoot,
} from "../utils/xast";
import { ArticleFrontMatter, DocInfo, Toc, TocItem } from "./types";

/**
 * Object to keep track of the global processing state during a transformation.
 * E.g., to generate unique ids, etc.
 */
export class PretextState {
    /** Information grabbed from the `<docinfo>` element. */
    docinfo: DocInfo;
    schema: JsonGrammar | null;
    _slugger: Slugger;
    _declaredIdCache: Set<string>;
    frontmatter: ArticleFrontMatter;
    /** Information for the table of contents. */
    toc: Toc;
    _tocInfoMap: Map<
        string,
        {
            numbering: number[];
            level: number;
            parent?: TocItem;
            item: TocItem;
        }
    >;
    root: XastRoot;
    /**
     * Function that continues processing content with XastToReact's `processContent` function.
     * This property must be set some time in the rendering phase.
     */
    processContent?: ProcessContentFunc;
    _parentMap: WeakMap<XastNode, XastNode>;
    _xastNodeToTocDivisionId: WeakMap<XastNode, string>;
    _xastDivisions: WeakSet<XastElement>;

    constructor(schema?: JsonGrammar) {
        this.docinfo = {};
        this.schema = schema || null;
        this._slugger = new Slugger();
        this._declaredIdCache = new Set();
        this.frontmatter = {
            titlepage: { authors: [], credits: [], editors: [] },
        };
        this.toc = [];
        this._tocInfoMap = new Map();
        this.root = { type: "root", children: [] };
        this._parentMap = new WeakMap();
        this._xastNodeToTocDivisionId = new WeakMap();
        this._xastDivisions = new WeakSet();
    }

    /**
     * Declares that `id` is an id that is explicitly declared in the document.
     * Returns `true` if the id has been declared already.
     */
    declareId(id: string) {
        if (
            this._declaredIdCache.has(id) ||
            this._slugger.occurrences[id] > 0
        ) {
            return true;
        }
        this._declaredIdCache.add(id);
        return false;
    }
    /**
     * Get an unused id that matches `preferredName`, if possible.
     */
    getUniqueId(preferredName?: string) {
        if (this._declaredIdCache.size > 0) {
            // If there are any ids in the cache, we need to drain it first.
            this._declaredIdCache.forEach((id) => this._slugger.slug(id, true));
            this._declaredIdCache.clear();
        }
        return this._slugger.slug(preferredName || "auto-generated-id", true);
    }

    setRoot(root: XastRoot) {
        this.root = root;
        this._generateParentMap();
        this._generateToc();
    }

    _generateParentMap() {
        visit(this.root, (node, info) => {
            const parent = info.parents[0];
            if (parent) {
                this._parentMap.set(node, parent);
            }
        });
    }

    /**
     * Generate a map from TOC item ids to information about that item.
     */
    _generateTocItemInfoMap() {
        const map: typeof this._tocInfoMap = new Map();

        const traverse = (items: TocItem[], level = 0, parent?: TocItem) => {
            for (let i = 0; i < items.length; i++) {
                const item = items[i];
                const parentInfo = map.get(parent?.id || "");
                let numbering = [i];
                if (parentInfo) {
                    numbering.unshift(...parentInfo.numbering);
                }

                map.set(item.id, {
                    level,
                    numbering,
                    item,
                    parent: parentInfo?.item,
                });

                traverse(item.children, level + 1, item);
            }
        };
        traverse(this.toc);
        this._tocInfoMap = map;
    }

    /**
     * Map every element to its nearest division parent.
     */
    _generateNodeToDivisionIdMap() {
        visit(this.root, (node, info) => {
            if (isElement(node) && this._xastDivisions.has(node)) {
                this._xastNodeToTocDivisionId.set(
                    node,
                    node.attributes?.["xml:id"] || ""
                );
                return;
            }
            // We just walk up the parents tree until we find one that is
            // recorded as a division. That is assumed to be its "division parent"
            for (const parent of info.parents) {
                if (isElement(parent) && this._xastDivisions.has(parent)) {
                    this._xastNodeToTocDivisionId.set(
                        node,
                        parent.attributes?.["xml:id"] || ""
                    );
                    break;
                }
            }
        });
    }

    /**
     * Extract table of contents information from the tree. Nodes are left intact in the process.
     */
    _generateToc() {
        const DIVISIONS = new Set([
            "book",
            "article",
            "part",
            "chapter",
            "section",
            "subsection",
            "subsubsection",
            "paragraphs",
            // These are special divisions; they can still be referenced.
            "frontmatter",
            "introduction",
            "conclusion",
            "headnote",
        ]);

        const isTitleNode = elmMatcher("title");
        const isDivision = (node: any): node is XastElement => {
            return isElement(node) && DIVISIONS.has(node.name);
        };

        const toc = this.toc;
        const tocItemsById: Record<string, TocItem> = {};

        visit(
            this.root,
            (node, info) => {
                const attrs = node.attributes || {};
                if (!attrs["xml:id"]) {
                    console.warn(
                        "Trying to add TOC entry, but item has not been assigned an `xml:id`. Did you forget to decorate all items with an id first?"
                    );
                    return;
                }
                const id = attrs["xml:id"];

                // Look for the title and try to make a new slug.
                const titleNode = node.children.find((node) =>
                    isTitleNode(node)
                );
                const title = isElement(titleNode)
                    ? onlyElementsAndText(titleNode.children)
                    : [];

                const newTocItem: TocItem = {
                    id,
                    title,
                    division: node.name,
                    children: [],
                };

                const parent = info.parents[0];
                if (!isElement(parent) || !parent.attributes?.["xml:id"]) {
                    // We're a root node with no immediate parent.
                    toc.push(newTocItem);
                } else {
                    const parentId = parent.attributes["xml:id"];
                    if (!tocItemsById[parentId]) {
                        throw new Error(
                            `Parent with id "${parentId}" not in the toc. Cannot insert child.`
                        );
                    }
                    tocItemsById[parentId].children.push(newTocItem);
                }
                tocItemsById[newTocItem.id] = newTocItem;

                // Keep track of all the nodes that we assigned a TOC item to.
                this._xastDivisions.add(node);
            },
            { test: isDivision }
        );

        this._generateTocItemInfoMap();
        this._generateNodeToDivisionIdMap();
    }

    /**
     * Get information about refable item that is included in the TOC.
     * The item is referred to by its id (explicit or auto-generated).
     */
    getTocItemInfo(node: XastNode) {
        const tocParentId = this._xastNodeToTocDivisionId.get(node);
        if (tocParentId) {
            return this._tocInfoMap.get(tocParentId);
        }
    }
}

// PartiallyRequired from https://pawelgrzybek.com/make-the-typescript-interface-partially-optional-required/
type PartiallyRequired<T, K extends keyof T> = Omit<T, K> &
    Required<Pick<T, K>>;
export type PretextStateWithProcessContent = PartiallyRequired<
    PretextState,
    "processContent"
>;
