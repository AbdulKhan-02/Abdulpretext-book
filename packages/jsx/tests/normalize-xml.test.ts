import util from "util";
import { normalizePretextXml } from "../src";
import { toString } from "hast-util-to-string";
import { toXml } from "xast-util-to-xml";
import { fromXml } from "xast-util-from-xml";
/* eslint-env jest */

// Make console.log pretty-print by default
const origLog = console.log;
console.log = (...args) => {
    origLog(...args.map((x) => util.inspect(x, false, 10, true)));
};

describe("Normalize PreTeXt XML", () => {
    it("Passes hello world", async () => {
        let source: string;
        source = `<pretext>
                    <article xml:id="hello-world">
                       <p>Hello, World!</p>
                    </article>
                </pretext>`;
        expect(toXml(normalizePretextXml(source))).toMatchSnapshot();
    });
});
