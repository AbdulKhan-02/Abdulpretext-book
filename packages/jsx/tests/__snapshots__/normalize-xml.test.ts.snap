// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Normalize PreTeXt XML Passes hello world 1`] = `"<pretext><article xml:id="hello-world"><p>Hello, World!</p></article></pretext>"`;
